src/Extensions/Compatability/Validation/Code.cs

```diff
@@ -0,0 +1,515 @@
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		/// <exception cref="ArgumentException">Thrown if any argument  <paramref name="argumentValue"/> element is null.</exception>
+		/// <exception cref="ArgumentNullException">Thrown if the supplied argument <paramref name="argumentValue"/> is null.</exception>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static IEnumerable<TEnumerable> ExpectsAllNotNull<TEnumerable>([ValidatedNotNull] IEnumerable<TEnumerable>? argumentValue, string argumentName, EventId? tagId)
+			where TEnumerable : class
+		{
+			argumentValue = ExpectsArgumentNotNull(argumentValue, argumentName, tagId);
+
+			if (!ValidateAllNotNull(argumentValue, argumentName, tagId))
+			{
+				ReportArgumentError(argumentName, AllErrorMessage);
+			}
+
+			return argumentValue;
+		}
+
+
+		/// <summary>
+		/// Checks the collection and throws an exception if it is null, contains no values, or contains any null values
+		/// </summary>
+		/// <typeparam name="T">Type of the collection</typeparam>
+		/// <param name="argumentValue">The argument value.</param>
+		/// <param name="argumentName">Name of the argument.</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed.</param>
+		/// <exception cref="ArgumentException">Thrown if <paramref name="argumentValue"/> is empty or if any element is null.</exception>
+		/// <exception cref="ArgumentNullException">Thrown if the supplied argument <paramref name="argumentValue"/> is null.</exception>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static IEnumerable<T> ExpectsNotEmptyAndAllNotNull<T>([ValidatedNotNull] ICollection<T>? argumentValue, string argumentName, EventId? tagId) where T : class
+		{
+			argumentValue = ExpectsArgumentNotNull(argumentValue, argumentName, tagId);
+
+			if (!ValidateNotEmptyAndAllNotNull(argumentValue, argumentName, tagId))
+			{
+				ReportArgumentError(argumentName, HasAnyErrorMessage);
+			}
+
+			return argumentValue;
+		}
+
+
+		/// <summary>
+		/// Checks the argument value and throws an exception if it is null or contains no values.
+		/// </summary>
+		/// <param name="argumentValue">The argument value.</param>
+		/// <param name="argumentName">Name of the argument.</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		/// <exception cref="ArgumentException">Thrown if the argument <paramref name="argumentValue"/> is empty.</exception>
+		/// <exception cref="ArgumentNullException">Thrown if the supplied argument <paramref name="argumentValue"/> is null.</exception>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static IEnumerable<T> ExpectsAny<T>([ValidatedNotNull] IEnumerable<T>? argumentValue, string argumentName, EventId? tagId)
+		{
+			argumentValue = ExpectsArgumentNotNull(argumentValue, argumentName, tagId);
+
+			if (!ValidateAny(argumentValue, argumentName, tagId))
```

Comment:
There's a lot of duplication throughout this class. Can you refactor the logic into shared helper functions and then use those?

Reply:
explained here https://github.com/microsoft/Omex/pull/153#discussion_r390509059